crates/circuit/src/dag_circuit.rs

```diff
@@ -2771,43 +2773,121 @@ def _format(operand):
+                                Interner::intern(&mut new_dag.cargs_cache, clbits.clone())?;
+
+                            let extr_attr = pi.extra_attrs.clone().unwrap();
+                            let new_pi = NodeType::Operation(PackedInstruction::new(
+                                pi.op.clone(),
+                                qubits_id,
+                                clbits_id,
+                                pi.params.clone(),
+                                extr_attr.label,
+                                extr_attr.duration,
+                                extr_attr.unit,
+                                extr_attr.condition,
+                                #[cfg(feature = "cache_pygates")]
+                                pi.py_op.borrow().clone(),
+                            ));
+                            let new_node = new_dag.dag.add_node(new_pi);
+                            node_map.insert(*node, new_node);
+                            non_classical = true;
+                        }
+                    },
+                    None => panic!("DAG node without payload!"),
+                }
+            }
+            if !non_classical {
+                continue;
+            }
+
+            // Handling the edges in the new dag
+            for node in comp_nodes {
+                // Since the nodes comprise an SCC, it's enough to just look at the (e.g.) outgoing edges
+                let outgoing_edges = self.dag.edges_directed(*node, Direction::Outgoing);
+
+                // Remove the edges added by copy_empty_like (as idle wires) to avoid duplication
+                if let Some(NodeType::QubitIn(_)) | Some(NodeType::ClbitIn(_)) =
+                    self.dag.node_weight(*node)
+                {
+                    let edges: Vec<EdgeIndex> = new_dag.dag.edges(*node).map(|e| e.id()).collect();
+                    for edge in edges {
+                        new_dag.dag.remove_edge(edge);
+                    }
+                }
+
+                for e in outgoing_edges {
+                    let (source, target) = (e.source(), e.target());
+                    let edge_weight = e.weight();
+                    let (source_new, target_new) = (
+                        node_map.get(&source).unwrap(),
+                        node_map.get(&target).unwrap(),
+                    );
+                    new_dag
+                        .dag
+                        .add_edge(*source_new, *target_new, edge_weight.clone());
+                }
+            }
+
+            if remove_idle_qubits {
+                let idle_wires: Vec<Bound<PyAny>> = new_dag
+                    .idle_wires(py, None)?
+                    .into_bound(py)
+                    .map(|q| q.unwrap())
+                    .filter(|e| {
+                        e.is_instance(new_dag.circuit_module.qubit.bind(py))
+                            .unwrap()
+                    })
+                    .collect();
```

Comment:
I think you'd be able to do this more optimally in your case by filtering the nodes in `new_dag.qubit_input_map` for any that don't have edges.

Also, `remove_qubits` would likely be failing because it would expect there to be a link between the input and output qubits (the idle wire you remove above). So perhaps you need to add that edge back in before calling.